### ISSUE TITLE
Enhancement: OpenAI Assistants API V2 support

### ISSUE THREAD

**Issue by danny-avila** [danny-avila/LibreChat#2490]
## Initial support added in https://github.com/danny-avila/LibreChat/pull/2781

> 5/19/24

Read the PR notes for more details on what changed, and what is still to be done.

---

### Discussed in https://github.com/danny-avila/LibreChat/discussions/2489

<div type='discussions-op-text'>

<sup>Originally posted by **sathyarajv** April 22, 2024</sup>
OpenAI updated the Assistants API to V2 with changes to the tools and files. Refer the Migration steps [here](https://platform.openai.com/docs/assistants/migration/what-has-changed)

![image](https://github.com/danny-avila/LibreChat/assets/9957258/e3717564-24a7-443e-99ea-2cb8565028dd)

Known issues: The attached files of an existing Assistants are not loading in the UI but its available on the OpenAI platform.  Migrating to V2 will resolve this issue.

Any plan on this migration?
</div>

**Comment by scottbetza** [comment]
Will this enhancement planned for v0.7.3 milestone completion allow for > 20 files?

**Comment by danny-avila** [comment]
> Will this enhancement planned for v0.7.3 milestone completion allow for > 20 files?

yes I think it's possible through vector stores

**Comment by danny-avila** [comment]
Due to several undocumented, breaking and non-breaking changes, as well as `gpt-4o` support being exclusive to v2, this is a priority item for this week.

**Comment by danny-avila** [comment]
Just a quick preview of what will be possible with Assistants V2: 


https://github.com/danny-avila/LibreChat/assets/110412045/20a42ba8-bfbf-498d-8914-f72d7b5d4d1c

**Comment by danny-avila** [comment]
I've made a lot of progress, and the initial support is almost done but not quite ready.

Azure integration breaks with the latest `openai` SDK versions, so I need to write old v1 methods for Azure file handling. Because of the differences with Azure, even later on, I've decided to make Azure Assistants its own endpoint, and there are also a few things to touch on to complete that process.

The way file attachments now work for tool resources is not fully complete for Code Interpreter, but is nearly done (just need to handle deleting from its tool_resources)

This is saying nothing of the new File Search vector store handling. This needs a completely new UI and file management system, which I've already started earlier this week.

My goal is to release the initial support tomorrow, with the ability to use all v2 features including file search, but no way manage/attach vector stores for file search yet, other than attaching files at the thread level.

I've addressed a lot of fixes for Assistants in the process, so I'm excited to get this out soon!

**Comment by danny-avila** [comment]
Merging soon, read the full notes if you're interested!

https://github.com/danny-avila/LibreChat/pull/2781

**Comment by danny-avila** [comment]
> @danny-avila thanks so much for your work on this. I saw that Azure OpenAI announced [support for assistants v2 today](https://learn.microsoft.com/en-us/azure/ai-services/openai/whats-new#assistants-v2-preview).
> 
> Passing this on in case it would (hopefully) be a comparatively easy task to add Azure support.

Cool! i'll see if it's as easy as flipping a switch at this point

**Comment by BennyMorton28** [comment]
Has anyone gotten gpt-4o working with Assistants in Librechat yet?

**Comment by danny-avila** [comment]
> Has anyone gotten gpt-4o working with Assistants in Librechat yet?

Yup. On the latest build, it's fully functional: 
![image](https://github.com/danny-avila/LibreChat/assets/110412045/c8643306-0691-4133-bbdc-89d64764dd77)

**Comment by BennyMorton28** [comment]
Sorry to keep troubling on this issue, I can't get it to work. I updated my assistant on OpenAI to be gpt-4o, and in my librechat.yaml file, and then when I rebuild I get the error below. I am using Docker, but I think there might be an update issue to the latest Librechat, but I'm not sure how to update/get the latest build? Thanks for any help

<img width="883" alt="Screenshot 2024-05-29 at 4 02 42 PM" src="https://github.com/danny-avila/LibreChat/assets/156146537/07fc9be4-54e1-48f4-a055-3e01d9c4c80f">

**Comment by danny-avila** [comment]
> Sorry to keep troubling on this issue, I can't get it to work. I updated my assistant on OpenAI to be gpt-4o, and in my librechat.yaml file, and then when I rebuild I get the error below. I am using Docker, but I think there might be an update issue to the latest Librechat, but I'm not sure how to update/get the latest build? Thanks for any help

I think you have a merge conflict while trying to update. Send an email to danny@librechat.ai and we can setup a time to troubleshoot live.

**Comment by danny-avila** [comment]
> I experience the same issue as the one above-mentioned. Is there any chance the error comes from wrong settings ni the librechat.yaml file ?

Only if its azure, the error above was not related to azure

**Comment by djuillard** [comment]
> > I experience the same issue as the one above-mentioned. Is there any chance the error comes from wrong settings ni the librechat.yaml file ?
> 
> Only if its azure, the error above was not related to azure

Yes, I'm on Azure.

**Comment by xfoggi** [comment]
OpenAI released GPT-4o-mini. Unfortunately, it runs only on V2. Any plans to migrate? :)
```
Error updating assistant 400 The requested model 'gpt-4o-mini' cannot be used with the Assistants API in v1. Follow the migration guide to upgrade to v2: https://platform.openai.com/docs/assistants/migration.
```

**Comment by kezzzsVS** [comment]
Just to add my voice on this, Azure OpenAI Assistants V2 support would be incredibly helpful. I would really like to be using GPT-4o for these (and soon GPT-4o mini)

**Comment by scottbetza** [comment]
Are vector stores now supported in v0.74?

**Comment by danny-avila** [comment]
> Are vector stores now supported in v0.74?

No, what you're asking for is a file browser for an OpenAI integration. Something like this will come for a native Agents integration before it's done for OpenAI, where you can already configure this on their dashboard

**Comment by rubenaster** [comment]
It would be really awesome to get Assistant API v2 support, especially since it also provides the code interpreter feature out-of-the-box.

It seems like the pull request is pretty close to completion.
